Add render stamp support

usakhelo opened the pull request:
Hi,

I added render stamp checkbox and an editbox in System rollout. Also TileCallbackFactory was added to work correctly with appleseed after-stamp callbacks.

Thanks for review!
Sergo.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/appleseedrendererparamdlg.cpp:
```diff
@@ -684,6 +684,7 @@ namespace
         ICustEdit*              m_text_renderingthreads;
         ISpinnerControl*        m_spinner_renderingthreads;
         AppleseedRenderer*      m_renderer;
+        ICustEdit*              m_text_render_stamp;
```
Keep data members ordered logically: `m_renderer` probably needs to stay at the end and `m_text_render_stamp` should stay close to the other settings.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/appleseedrendererparamdlg.cpp:
```diff
@@ -706,11 +707,15 @@ namespace
 
         virtual void init(HWND hwnd) override
         {
+            m_text_render_stamp = GetICustEdit(GetDlgItem(hwnd, IDC_TEXT_RENDER_STAMP));
```
Same thing here, keep initialization order consistent with the settings order, data member orders, etc.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/tilecallback.cpp:
```diff
@@ -264,3 +264,26 @@ void TileCallback::blit_tile(
+
+asr::ITileCallback* TileCallbackFactory::create()
+{
+    return new TileCallback(m_bitmap, m_rendered_tile_count);
```
That may be incorrect.

There was a reason why we were passing a single tile callback instead of a factory: It's likely that Max doesn't support updating the rendered bitmap or the window from multiple threads, which is exactly what's going to happen with this change.

I don't know if Max supports that. If it doesn't, then that tile callback factory should always return the same tile callback instance (and make sure it only gets destroyed once of course).

usakhelo replied:
I think it's better to update bitmap from single thread.
But how returning single instance of TileCallback prevent appleseed from running it from multiple threads?

dictoon replied:
You're of course right, that was stupid of me :) We would need to enqueue the requests and run them from the main thread... Which is exactly what the `SerialTileCallback` does.

It seems to me that maybe the right solution would be to continue using a single tilecallback (instead of a factory), and fix `MasterRenderer` in the sense that it should execute all pending callback updates after adding the render stamp.

usakhelo commented:
Hi @dictoon,
I've checked recent changes to appleseed and it seems they fixed the issue. So I'm rolling back all the changes to TileCallback.

usakhelo commented:
Also addressing the issues you've noted.
About saving `m_log_open_mode` and `m_log_material_editor_messages` - these settings are saved into system ini file which makes those settings global across all the scenes.

dictoon commented:
> I've checked recent changes to appleseed and it seems they fixed the issue. So I'm rolling back all the changes to TileCallback.

Good to hear that they fix the issue! Thanks.

> About saving m_log_open_mode and m_log_material_editor_messages - these settings are saved into system ini file which makes those settings global across all the scenes.

Ok, that's perfect, thanks for the clarification.

usakhelo commented:
I've noticed in corona renderer stamp format editbox is not disabled when stamp is off. It seams reasonable. I'll just leave this here as a reminder:
![image](https://user-images.githubusercontent.com/1850877/39101574-19f27ecc-4669-11e8-8a05-370c3fa5649d.png)

dictoon commented:
After further discussion, this appears safe to merge so here we go.
